Kernel/Storage/SD/SDHostController.cpp

```diff
@@ -0,0 +1,671 @@
+            StorageDevice::LUNAddress { controller_id(), 0, 0 },
+            hardware_relative_controller_id(), BLOCK_LEN,
+            card_capacity_in_blocks, rca, ocr, cid, scr)));
+}
+
+bool SDHostController::retry_with_timeout(Function<bool()> f,
+    i64 delay_between_tries)
+{
+    int timeout = 1000;
+    bool success = false;
+    while (!success && timeout > 0) {
+        success = f();
+        if (!success)
+            delay(delay_between_tries);
+        timeout--;
+    }
+    return timeout > 0;
+}
+
+ErrorOr<void> SDHostController::issue_command(SD::CommandIndex index,
+    u32 argument)
+{
+    // SDHC 3.7.1 Transaction Control without Data Transfer Using DAT Line
+    constexpr u32 COMMAND_INHIBIT = 1 << 1;
+    auto cmd = SD::get_command(index);
+
+    // 1. Check Command Inhibit (CMD) in the Present State register.
+    //    Repeat this step until **Command Inhibit (CMD)** is 0.
+    //    That is, when Command Inhibit (CMD) is 1, the Host Driver
+    //    shall not issue an SD Command.
+    if (!retry_with_timeout(
+            [&]() { return !(m_registers->present_state & COMMAND_INHIBIT); })) {
+        return EIO;
+    }
+
+    // 2. If the Host Driver issues an SD Command using DAT lines
+    //    including busy signal, go to step (3).
+    //    If without using DAT lines including busy signal, go to step (5).
+    // 3. If the Host Driver is issuing an abort command, go to step (5). In the
+    // case of non-abort
+    //    command, go to step (4).
+    if (cmd.requires_dat_line() && !cmd.is_abort()) {
+
+        // 4. Check Command Inhibit (DAT) in the Present State register. Repeat
+        // this step until
+        //    Command Inhibit (DAT) is set to 0.
+        constexpr u32 DATA_INHIBIT = 1 << 2;
+        if (!retry_with_timeout(
+                [&]() { return !(m_registers->present_state & DATA_INHIBIT); })) {
+            return EIO;
+        }
+    }
+
+    // 5. Set registers as described in Table 1-2 except Command register.
+    m_registers->argument_1 = argument;
+
+    // 6. Set the Command register.
+    m_registers->transfer_mode_and_command = cmd.to_u32();
+
+    // 7. Perform Command Completion Sequence in accordance with 3.7.1.2.
+    // Done in wait_for_response()
+
+    return {};
+}
+
+ErrorOr<SDHostController::Response> SDHostController::wait_for_response()
+{
+    // SDHC 3.7.1.2 The Sequence to Finalize a Command
+
+    // 1. Wait for the Command Complete Interrupt. If the Command Complete
+    // Interrupt has occurred,
+    //    go to step (2).
+    if (!retry_with_timeout(
+            [&]() { return m_registers->interrupt_status & COMMAND_COMPLETE; })) {
+        return EIO;
+    }
+
+    // 2. Write 1 to Command Complete in the Normal Interrupt Status register to
+    // clear this bit
+    m_registers->interrupt_status = COMMAND_COMPLETE;
+
+    // 3. Read the Response register(s) to get the response.
+    struct Response r = {};
```

Comment:
no need for the `struct` specifier
(I think you used that in other places as well)